File: src/gamification.py
import json
import os
from utils import get_external_path

class UserProfile:
    def __init__(self):
        self.xp = 0
        self.level = 1
        self.stats = {
            "total_correct": 0,
            "total_attempts": 0,
            "current_streak": 0,
            "best_streak": 0
        }
        self.achievements = []
        self.filepath = get_external_path(os.path.join("data", "user_progress.json"))
        self.load()

    def load(self):
        if os.path.exists(self.filepath):
            try:
                with open(self.filepath, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    self.xp = data.get("xp", 0)
                    self.level = data.get("level", 1)
                    self.stats = data.get("stats", self.stats)
                    self.achievements = data.get("achievements", [])
            except Exception as e:
                print(f"Error loading user profile: {e}")

    def save(self):
        data = {
            "xp": self.xp,
            "level": self.level,
            "stats": self.stats,
            "achievements": self.achievements
        }
        try:
            # Ensure directory exists
            os.makedirs(os.path.dirname(self.filepath), exist_ok=True)
            with open(self.filepath, "w", encoding="utf-8") as f:
                json.dump(data, f, indent=4)
        except Exception as e:
            print(f"Error saving user profile: {e}")

    def add_xp(self, amount):
        old_level = self.level
        self.xp += amount
        self.calculate_level()
        return self.level > old_level

    def calculate_level(self):
        # Formula: Level = 1 + (XP // 100)
        # Linear progression: 100 XP per level
        self.level = 1 + (self.xp // 100)

    def record_attempt(self, correct: bool):
        self.stats["total_attempts"] += 1
        if correct:
            self.stats["total_correct"] += 1
            self.stats["current_streak"] += 1
            if self.stats["current_streak"] > self.stats["best_streak"]:
                self.stats["best_streak"] = self.stats["current_streak"]
        else:
            self.stats["current_streak"] = 0

        self.check_achievements()
        self.save()

    def check_achievements(self):
        new_unlocks = []

        # Define Achievements conditions
        # (ID, Name, Description, Condition Lambda)
        definitions = [
            ("first_step", "Primer Paso", "Responde correctamente tu primera pregunta",
             lambda s: s["total_correct"] >= 1),
            ("scholar", "Estudioso", "Responde 10 preguntas correctamente",
             lambda s: s["total_correct"] >= 10),
            ("master", "Maestro", "Responde 50 preguntas correctamente",
             lambda s: s["total_correct"] >= 50),
            ("on_fire", "En Llamas", "Consigue una racha de 5 aciertos seguidos",
             lambda s: s["current_streak"] >= 5),
            ("unstoppable", "Imparable", "Consigue una racha de 10 aciertos seguidos",
             lambda s: s["current_streak"] >= 10)
        ]

        for aid, name, desc, condition in definitions:
            if aid not in self.achievements:
                if condition(self.stats):
                    self.achievements.append(aid)
                    new_unlocks.append((name, desc))

        return new_unlocks

    def get_progress_to_next_level(self):
        # Returns (current_level_xp, xp_needed_for_next_level)
        # Since level is every 100 XP
        # XP = 150. Level = 2. XP into level = 50.
        # XP needed = 100.
        current_level_xp = self.xp % 100
        return current_level_xp, 100

File: desktop-app/src/lib/gamification.ts
/**
 * Interfaz que define las estadísticas del usuario.
 */
export interface UserStats {
    /** Número total de respuestas correctas. */
    total_correct: number;
    /** Número total de intentos realizados. */
    total_attempts: number;
    /** Racha actual de respuestas correctas consecutivas. */
    current_streak: number;
    /** Mejor racha histórica de respuestas correctas. */
    best_streak: number;
}

/**
 * Interfaz que define un logro desbloqueado.
 */
export interface Achievement {
    /** Identificador único del logro. */
    id: string;
    /** Nombre del logro. */
    name: string;
    /** Descripción del logro. */
    description: string;
    /** Fecha de desbloqueo (opcional). */
    unlockedAt?: string;
}

const STORAGE_KEY = 'physicode_user_profile';

/**
 * Clase que gestiona el perfil del usuario, incluyendo experiencia, nivel, estadísticas y logros.
 * Utiliza localStorage para persistencia de datos.
 */
export class UserProfile {
    /** Puntos de experiencia acumulados. */
    xp: number = 0;
    /** Gemas o moneda virtual acumulada. */
    gems: number = 0;
    /** Nivel actual del usuario. */
    level: number = 1;
    /** Estadísticas de rendimiento del usuario. */
    stats: UserStats = {
        total_correct: 0,
        total_attempts: 0,
        current_streak: 0,
        best_streak: 0,
    };
    /** Lista de IDs de logros desbloqueados. */
    achievements: string[] = [];

    /**
     * Inicializa el perfil de usuario cargando los datos guardados.
     */
    constructor() {
        this.load();
    }

    /**
     * Carga los datos del perfil desde localStorage.
     * @private
     */
    private load() {
        try {
            const data = localStorage.getItem(STORAGE_KEY);
            if (data) {
                const parsed = JSON.parse(data);
                this.xp = parsed.xp || 0;
                this.gems = parsed.gems || 0;
                this.level = parsed.level || 1;
                this.stats = { ...this.stats, ...parsed.stats };
                this.achievements = parsed.achievements || [];
            }
        } catch (e) {
            console.error("Failed to load user profile", e);
        }
    }

    /**
     * Guarda el estado actual del perfil en localStorage y emite un evento de actualización.
     */
    save() {
        try {
            const data = {
                xp: this.xp,
                gems: this.gems,
                level: this.level,
                stats: this.stats,
                achievements: this.achievements,
            };
            localStorage.setItem(STORAGE_KEY, JSON.stringify(data));
            // Dispatch event for UI updates
            window.dispatchEvent(new Event('profile-updated'));
        } catch (e) {
            console.error("Failed to save user profile", e);
        }
    }

    /**
     * Añade experiencia al usuario y calcula si ha subido de nivel.
     * @param amount Cantidad de experiencia a añadir.
     * @returns {boolean} True si el usuario subió de nivel, False en caso contrario.
     */
    addXp(amount: number): boolean {
        const oldLevel = this.level;
        this.xp += amount;
        this.calculateLevel();
        this.save();
        return this.level > oldLevel;
    }

    /**
     * Añade gemas al usuario.
     * @param amount Cantidad de gemas a añadir.
     */
    addGems(amount: number) {
        this.gems += amount;
        this.save();
    }

    /**
     * Calcula el nivel actual basado en la experiencia total.
     * Progresión lineal: 100 XP por nivel.
     */
    calculateLevel() {
        // Linear progression: 100 XP per level
        this.level = 1 + Math.floor(this.xp / 100);
    }

    /**
     * Registra un intento de respuesta (correcta o incorrecta) y actualiza estadísticas.
     * @param correct Indica si la respuesta fue correcta.
     * @returns {{ leveledUp: boolean; newAchievements: Achievement[] }} Objeto indicando si hubo subida de nivel y nuevos logros.
     */
    recordAttempt(correct: boolean): { leveledUp: boolean; newAchievements: Achievement[] } {
        this.stats.total_attempts += 1;
        let leveledUp = false;

        if (correct) {
            this.stats.total_correct += 1;
            this.stats.current_streak += 1;
            if (this.stats.current_streak > this.stats.best_streak) {
                this.stats.best_streak = this.stats.current_streak;
            }
            // Award XP for correct answer
            leveledUp = this.addXp(10);
        } else {
            this.stats.current_streak = 0;
        }

        const newAchievements = this.checkAchievements();
        this.save();

        return { leveledUp, newAchievements };
    }

    /**
     * Verifica si se han cumplido las condiciones para desbloquear nuevos logros.
     * @returns {Achievement[]} Lista de logros recién desbloqueados.
     */
    checkAchievements(): Achievement[] {
        const unlocked: Achievement[] = [];

        const definitions = [
            { id: "first_step", name: "Primer Paso", description: "Responde correctamente tu primera pregunta", condition: (s: UserStats) => s.total_correct >= 1 },
            { id: "scholar", name: "Estudioso", description: "Responde 10 preguntas correctamente", condition: (s: UserStats) => s.total_correct >= 10 },
            { id: "master", name: "Maestro", description: "Responde 50 preguntas correctamente", condition: (s: UserStats) => s.total_correct >= 50 },
            { id: "on_fire", name: "En Llamas", description: "Consigue una racha de 5 aciertos seguidos", condition: (s: UserStats) => s.current_streak >= 5 },
            { id: "unstoppable", name: "Imparable", description: "Consigue una racha de 10 aciertos seguidos", condition: (s: UserStats) => s.current_streak >= 10 },
        ];

        for (const def of definitions) {
            if (!this.achievements.includes(def.id) && def.condition(this.stats)) {
                this.achievements.push(def.id);
                unlocked.push({ id: def.id, name: def.name, description: def.description });
            }
        }

        return unlocked;
    }

    /**
     * Obtiene el progreso hacia el siguiente nivel.
     * @returns {{ current: number; max: number }} Objeto con XP actual en el nivel y XP requerida para el siguiente.
     */
    getNextLevelProgress() {
        const currentLevelXp = this.xp % 100;
        return { current: currentLevelXp, max: 100 };
    }
}

/** Instancia global del perfil de usuario. */
export const userProfile = new UserProfile();
